### PR TITLE
[PYTHON] Optional Cython Module for Symbols

### PR DESCRIPTION
**Pull request by tqchen**
Only performance critical parts are done using cython, most codes go through ctypes

- the folder cython contains cython specific implementation of certain modules
   - cython class is compatible with ctypes class, so other slow module can simply be implemented by ctypes without using cython.
- the folder ctypes contains ctypes specific implementation of certain modules

When cython is not available, ctypes will be used by default. The build package will work under any python version and regardless whether cython module is available

cython module can be build via

```make cython2``` and ```make cython3```.

### CODE REVIEW

**Comment by tqchen** [comment]
TODO, possibly in another PR, move more logics to common code, specifically, support the symbol_doc.py doc build mechanism.

File: Makefile [hunk 1]
```diff
@@ -93,7 +93,8 @@ ifeq ($(USE_DIST_KVSTORE), 1)
 	LDFLAGS += $(PS_LDFLAGS_A)
 endif
 
-.PHONY: clean all test lint doc clean_all rcpplint rcppexport roxygen
+.PHONY: clean all test lint doc clean_all rcpplint rcppexport roxygen\
+	cython2 cython3 cyclean
```
Comment (piiswrong) [inline Makefile]: clean together with clean

File: python/mxnet/cython/symbol.pyx [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+        Parameters
+        ----------
+        **kwargs
+            The attributes to set
+        """
+        SymbolSetAttr(self.chandle, kwargs)
+
+    def __reduce__(self):
+        return (_symbol_cls, (None,), self.__getstate__())
+
```
Comment (piiswrong) [inline python/mxnet/cython/symbol.pyx]: missing _compose?

Comment (tqchen) [inline python/mxnet/cython/symbol.pyx]: I decided that compose can in slow path and moved it to common area, can be moved back later

**Comment by tqchen** [comment]
I am merging this in, to allow next PR of FCompute support to come up, have tested the lint and testcases locally.

**Comment by tqchen** [comment]
https://github.com/dmlc/mxnet/pull/3249

**Pull request merged**